File: prisma/seed.ts
// tsx no lee .env por su cuenta: sin esto habría que exportar
// DATABASE_URL y SEED_OWNER_PASSWORD a mano en cada ejecución.
import "dotenv/config";
import { PrismaClient, Role } from "@prisma/client";
import bcrypt from "bcryptjs";

const prisma = new PrismaClient();

const products = [
  {
    sku: "KN-SF1",
    type: "ABANICO",
    name: "Abanico de pedestal",
    description: 'Tamaño 18"',
    color: "NEGRO",
    listPrice: 2525,
    discountPct: 10,
    netPrice: 2272.5,
    stockOnHand: 188,
  },
  {
    sku: "KN-0151",
    type: "ESTUFA",
    name: "Estufa de 20'' Satinada",
    description: "Tamaño 20''",
    color: "SATINADA",
    listPrice: 10965,
    discountPct: 0,
    netPrice: 10965,
    stockOnHand: 12,
  },
  {
    sku: "KN-76N",
    type: "ESTUFA",
    name: "Estufa de 30'' Satinada",
    description: "Tamaño 30''",
    color: "SATINADA",
    listPrice: 24995,
    discountPct: 0,
    netPrice: 24995,
    stockOnHand: 185,
  },
  {
    sku: "KN-76G",
    type: "ESTUFA",
    name: 'Estufa de 30" tope en cristal',
    description: 'Tamaño 30"',
    color: "SATINADA",
    listPrice: 23750,
    discountPct: 0,
    netPrice: 23750,
    stockOnHand: 273,
  },
  {
    sku: "KN-6201",
    type: "ESTUFA",
    name: 'Estufa de 24" libras',
    description: 'Tamaño 24"',
    color: "SATINADA",
    listPrice: 14750,
    discountPct: 0,
    netPrice: 14750,
    stockOnHand: 148,
  },
  {
    sku: "KN-7600",
    type: "ESTUFA",
    name: 'Estufa de 30"',
    description: 'Tamaño 30"',
    color: "SATINADA",
    listPrice: 19455,
    discountPct: 0,
    netPrice: 19455,
    stockOnHand: 322,
  },
  {
    sku: "KN-60X",
    type: "ESTUFA",
    name: 'Estufa 24" Alta Gama',
    description: 'Tamaño 24"',
    color: "SATINADA",
    listPrice: 25775,
    discountPct: 0,
    netPrice: 25775,
    stockOnHand: 94,
  },
  {
    sku: "KN-80X",
    type: "ESTUFA",
    name: 'Estufa 30" Alta Gama',
    description: 'Tamaño 30"',
    color: "SATINADA",
    listPrice: 31395,
    discountPct: 0,
    netPrice: 31395,
    stockOnHand: 0,
  },
  {
    sku: "KN-18",
    type: "LAVADORA",
    name: "Lavadora 18 libras",
    description: "Tamaño 18 libras",
    color: null,
    listPrice: 9485,
    discountPct: 0,
    netPrice: 9485,
    stockOnHand: 29,
  },
  {
    sku: "KN-22",
    type: "LAVADORA",
    name: "Lavadora 22 libras",
    description: "Tamaño 22 libras",
    color: null,
    listPrice: 11240,
    discountPct: 0,
    netPrice: 11240,
    stockOnHand: 56,
  },
  {
    sku: "KN-30",
    type: "LAVADORA",
    name: "Lavadora 30 libras",
    description: "Tamaño 30 libras",
    color: null,
    listPrice: 12860,
    discountPct: 0,
    netPrice: 12860,
    stockOnHand: 15,
  },
  {
    sku: "KN-560",
    type: "NEVERA",
    name: "Nevera 2 puertas",
    description: "Tamaño 20 pies",
    color: "SATINADA",
    listPrice: 55940,
    discountPct: 3,
    netPrice: 54261.8,
    stockOnHand: 43,
  },
];

async function main() {
  // La clave venía escrita aquí, y además publicada en el README. Quitarla
  // del formulario de login no servía de nada mientras el seed la siguiera
  // creando sola en cada despliegue.
  const password = process.env.SEED_OWNER_PASSWORD;
  if (!password || password.length < 8) {
    throw new Error(
      "Define SEED_OWNER_PASSWORD (mínimo 8 caracteres) antes de sembrar. Ejemplo:\n" +
        '  SEED_OWNER_PASSWORD="tu-clave-segura" npm run db:seed'
    );
  }
  const passwordHash = await bcrypt.hash(password, 10);

  await prisma.user.upsert({
    where: { email: "dueno@kngold.com.do" },
    update: {},
    create: {
      name: "Dueño KN GOLD",
      email: "dueno@kngold.com.do",
      passwordHash,
      role: Role.OWNER,
    },
  });

  await prisma.user.upsert({
    where: { email: "vendedor@kngold.com.do" },
    update: {},
    create: {
      name: "Vendedor Demo",
      email: "vendedor@kngold.com.do",
      passwordHash,
      role: Role.SELLER,
    },
  });

  await prisma.user.upsert({
    where: { email: "vendedor2@kngold.com.do" },
    update: {},
    create: {
      name: "Vendedor Norte",
      email: "vendedor2@kngold.com.do",
      passwordHash,
      role: Role.SELLER,
    },
  });

  for (const p of products) {
    await prisma.product.upsert({
      where: { sku: p.sku },
      update: {
        name: p.name,
        type: p.type,
        description: p.description,
        color: p.color,
        listPrice: p.listPrice,
        discountPct: p.discountPct,
        netPrice: p.netPrice,
        stockOnHand: p.stockOnHand,
      },
      create: p,
    });
  }

  await prisma.appSetting.upsert({
    where: { key: "reservation_hours" },
    update: { value: "48" },
    create: { key: "reservation_hours", value: "48" },
  });

  console.log(
    "Seed OK — dueno@kngold.com.do (dueño), vendedor@kngold.com.do y " +
      "vendedor2@kngold.com.do (vendedores). Clave: la de SEED_OWNER_PASSWORD."
  );
}

main()
  .catch((e) => {
    console.error(e);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });

File: backend/internal/db/seed.go
package db

import (
	"database/sql"
	"time"

	"github.com/google/uuid"
	"github.com/kngold/sistema/internal/auth"
)

type productSeed struct {
	SKU, Type, Name, Description, Color string
	ListPrice, DiscountPct, NetPrice    float64
	Stock                               int
}

func Seed(db *sql.DB) error {
	var n int
	if err := db.QueryRow(`SELECT COUNT(*) FROM users`).Scan(&n); err != nil {
		return err
	}
	if n > 0 {
		return nil
	}

	hash, err := auth.HashPassword("kngold2026")
	if err != nil {
		return err
	}
	now := time.Now().UTC().Format(time.RFC3339)

	users := []struct {
		name, email, role string
	}{
		{"Dueño KN GOLD", "dueno@kngold.com.do", "OWNER"},
		{"Vendedor Demo", "vendedor@kngold.com.do", "SELLER"},
		{"Vendedor Norte", "vendedor2@kngold.com.do", "SELLER"},
	}
	for _, u := range users {
		_, err := db.Exec(`
			INSERT INTO users (id, name, email, password_hash, role, active, created_at, updated_at)
			VALUES (?, ?, ?, ?, ?, 1, ?, ?)`,
			uuid.NewString(), u.name, u.email, hash, u.role, now, now)
		if err != nil {
			return err
		}
	}

	products := []productSeed{
		{"KN-SF1", "ABANICO", "Abanico de pedestal", `Tamaño 18"`, "NEGRO", 2525, 10, 2272.5, 188},
		{"KN-0151", "ESTUFA", "Estufa de 20'' Satinada", "Tamaño 20''", "SATINADA", 10965, 0, 10965, 12},
		{"KN-76N", "ESTUFA", "Estufa de 30'' Satinada", "Tamaño 30''", "SATINADA", 24995, 0, 24995, 185},
		{"KN-76G", "ESTUFA", `Estufa de 30" tope en cristal`, `Tamaño 30"`, "SATINADA", 23750, 0, 23750, 273},
		{"KN-6201", "ESTUFA", `Estufa de 24" libras`, `Tamaño 24"`, "SATINADA", 14750, 0, 14750, 148},
		{"KN-7600", "ESTUFA", `Estufa de 30"`, `Tamaño 30"`, "SATINADA", 19455, 0, 19455, 322},
		{"KN-60X", "ESTUFA", `Estufa 24" Alta Gama`, `Tamaño 24"`, "SATINADA", 25775, 0, 25775, 94},
		{"KN-80X", "ESTUFA", `Estufa 30" Alta Gama`, `Tamaño 30"`, "SATINADA", 31395, 0, 31395, 0},
		{"KN-18", "LAVADORA", "Lavadora 18 libras", "Tamaño 18 libras", "", 9485, 0, 9485, 29},
		{"KN-22", "LAVADORA", "Lavadora 22 libras", "Tamaño 22 libras", "", 11240, 0, 11240, 56},
		{"KN-30", "LAVADORA", "Lavadora 30 libras", "Tamaño 30 libras", "", 12860, 0, 12860, 15},
		{"KN-560", "NEVERA", "Nevera 2 puertas", "Tamaño 20 pies", "SATINADA", 55940, 3, 54261.8, 43},
	}
	for _, p := range products {
		_, err := db.Exec(`
			INSERT INTO products (
				id, sku, name, type, description, color, list_price, discount_pct, net_price,
				stock_on_hand, active, created_at, updated_at
			) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 1, ?, ?)`,
			uuid.NewString(), p.SKU, p.Name, p.Type, nullEmpty(p.Description), nullEmpty(p.Color),
			p.ListPrice, p.DiscountPct, p.NetPrice, p.Stock, now, now)
		if err != nil {
			return err
		}
	}

	_, err = db.Exec(`
		INSERT INTO app_settings (id, key, value) VALUES (?, 'reservation_hours', '48')`,
		uuid.NewString())
	return err
}

func nullEmpty(s string) any {
	if s == "" {
		return nil
	}
	return s
}
